File: Frontend/BibliWorm/src/app/core/services/department.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { CreateUpdateDepartmentModel } from '../models/department/create-update-department-model';
import { Observable } from 'rxjs';
import { api, extractData } from '../constants/api-constants';
import { DepartmentModel } from '../models/department/department-model';

@Injectable({
    providedIn: 'root'
})

export class DepartmentService {

    constructor(
        private http: HttpClient,
    ) { }

    public addDepartment(departmentModel: CreateUpdateDepartmentModel): Observable<any> {
        var url = api.department.addDepartment;

        return this.http.post(
            url, 
            departmentModel
        ).pipe(extractData());
    }

    public deleteDepartment(departmentId: number): Observable<any> {
        var url = api.department.deleteDepartment;
        var params = new HttpParams().set('departmentId', departmentId);

        return this.http.delete(
            url,
            { params: params }
        ).pipe(extractData());
    }

    public getAllDepartments(): Observable<DepartmentModel[]> {
        var url = api.department.getAllDepartments;

        return this.http.get(url)
            .pipe(extractData<DepartmentModel[]>());
    }

    public getDepartmentById(departmentId: number): Observable<DepartmentModel> {
        var url = api.department.getDepartmentById;
        var params = new HttpParams().set('departmentId', departmentId);

        return this.http.get(
            url,
            { params: params }
        ).pipe(extractData<DepartmentModel>())
    }

    public updateDepartment(departmentModel: CreateUpdateDepartmentModel): Observable<any> {
        var url = api.department.updateDepartment;
        
        return this.http.post(
            url,
            departmentModel
        ).pipe(extractData());
    }
}

File: Backend/BLL/Services/DepartmentService.cs
using AutoMapper;
using BLL.Contracts;
using BLL.Infrastructure.Models.Department;
using DAL.Contracts;
using Domain.Models;
using Microsoft.EntityFrameworkCore;

namespace BLL.Services;
public class DepartmentService : IDepartmentService
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly Lazy<IMapper> _mapper;

    public DepartmentService(
        IUnitOfWork unitOfWork,
        Lazy<IMapper> mapper)
    {
        _unitOfWork = unitOfWork;
        _mapper = mapper;
    }

    public void AddDepartment(CreateUpdateDepartmentModel departmentModel)
    {
        var department = _mapper.Value.Map<Department>(departmentModel);

        _unitOfWork.Departments.Value.Create(department);
    }

    public void AddStatisticsForDepartment(CreateDepartmentStatisticsModel departmentStatisticsModel)
    {
        var departmentStatistics = _mapper.Value.Map<DepartmentStatistics>(departmentStatisticsModel);

        _unitOfWork.Departments.Value.CreateDepartmentStatistics(departmentStatistics);
    }

    public void DeleteDepartment(int departmentId)
    {
        _unitOfWork.Departments.Value.Delete(departmentId);
    }

    public IEnumerable<DepartmentModel> GetAllDepartments()
    {
        var departments = _unitOfWork.Departments.Value.GetAll()
            .Include(d => d.DepartmentStatistics);
        var departmentModels = _mapper.Value.Map<List<DepartmentModel>>(departments);

        return departmentModels;
    }

    public DepartmentModel GetDepartmentById(int departmentId)
    {
        var department = _unitOfWork.Departments.Value.GetById(departmentId);
        var departmentModel = _mapper.Value.Map<DepartmentModel>(department);

        return departmentModel;
    }

    public void UpdateDepartment(CreateUpdateDepartmentModel departmentModel)
    {
        var department = _mapper.Value.Map<Department>(departmentModel);

        _unitOfWork.Departments.Value.Update(department);
    }
}
